src/main/cc/common_cpp/protobuf_util/riegeli_io.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_MAIN_CC_COMMON_CPP_PROTOBUF_UTIL_RIEGELI_IO_H_
+#define SRC_MAIN_CC_COMMON_CPP_PROTOBUF_UTIL_RIEGELI_IO_H_
+
+#include <fcntl.h>
+#include <stdio.h>
+
+#include <vector>
+
+#include "absl/status/status.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "riegeli/bytes/fd_reader.h"
+#include "riegeli/bytes/fd_writer.h"
+#include "riegeli/records/record_reader.h"
+#include "riegeli/records/record_writer.h"
+
+namespace wfa {
+
+using ::riegeli::FdReader;
+using ::riegeli::FdWriter;
+using ::riegeli::RecordReader;
+using ::riegeli::RecordWriter;
+
+// Read a list of protobuf messages to @messages. The input messages are from
+// the file with @path.
+// Return error status if the reading fails.
+// The details about the format of the input file can be found in
+// https://github.com/google/riegeli/blob/master/doc/riegeli_records_file_format.md
+template <typename Message>
+absl::Status ReadRiegeliFile(absl::string_view path,
+                             std::vector<Message>& messages) {
+  RecordReader reader(FdReader(path, O_RDONLY));
+  messages.clear();
+  while (true) {
+    Message message;
+    if (!reader.ReadRecord(message)) {
+      if (reader.healthy()) {
+        // EOF
+        break;
+      } else {
+        return absl::InvalidArgumentError(
+            absl::StrCat("Unable to read message from file: ", path));
+      }
+    }
+    messages.emplace_back(message);
```

Comment:
Move instead of copying: `messages.push_back(std::move(message))`

An alternative which avoids even the moves:

```
while (reader.ReadRecord(messages.emplace_back())) {}
messages.pop_back();
if (!reader.Close()) {
  return reader.status();
}
return absl::OkStatus();
```